lib/astutils.cpp

```diff
@@ -1210,114 +1210,107 @@ SmallVector<ReferenceToken> followAllReferences(const Token* tok,
-        std::set<ReferenceToken, ReferenceTokenLess> result;
-        std::vector<const Token*> returns = Function::findReturns(f);
-        for (const Token* returnTok : returns) {
-            if (returnTok == tok)
-                continue;
-            for (const ReferenceToken& rt :
-                 followAllReferences(returnTok, temporary, inconclusive, errors, depth - returns.size())) {
-                const Variable* argvar = rt.token->variable();
-                if (!argvar) {
-                    refs_result.push_back({tok, std::move(errors)});
-                    return refs_result;
-                }
-                if (argvar->isArgument() && (argvar->isReference() || argvar->isRValueReference())) {
-                    const int n = getArgumentPos(argvar, f);
-                    if (n < 0) {
-                        refs_result.push_back({tok, std::move(errors)});
-                        return refs_result;
-                    }
-                    std::vector<const Token*> args = getArguments(tok->previous());
-                    if (n >= args.size()) {
-                        refs_result.push_back({tok, std::move(errors)});
-                        return refs_result;
+        } else if (tok->previous() && tok->previous()->function() && Token::Match(tok->previous(), "%name% (")) {
+            const Function *f = tok->previous()->function();
+            if (!Function::returnsReference(f)) {
+                break;
+            }
+            std::set<ReferenceToken, ReferenceTokenLess> result;
+            std::vector<const Token*> returns = Function::findReturns(f);
+            for (const Token* returnTok : returns) {
+                if (returnTok == tok)
+                    continue;
+                for (const ReferenceToken& rt :
+                     followAllReferences(returnTok, temporary, inconclusive, errors, depth - returns.size())) {
+                    const Variable* argvar = rt.token->variable();
+                    if (!argvar) {
+                        break;
                     }
-                    const Token* argTok = args[n];
-                    ErrorPath er = errors;
-                    er.emplace_back(returnTok, "Return reference.");
-                    er.emplace_back(tok->previous(), "Called function passing '" + argTok->expressionString() + "'.");
-                    auto refs =
-                        followAllReferences(argTok, temporary, inconclusive, std::move(er), depth - returns.size());
-                    result.insert(refs.cbegin(), refs.cend());
-                    if (!inconclusive && result.size() > 1) {
-                        refs_result.push_back({tok, std::move(errors)});
-                        return refs_result;
+                    if (argvar->isArgument() && (argvar->isReference() || argvar->isRValueReference())) {
+                        const int n = getArgumentPos(argvar, f);
+                        if (n < 0) {
+                            break;
```

Comment:
This changes the behavior. Instead of exiting function this will only exit the `for` loop.

Reply:
That is intentional. It exits the loop, `result` is empty and then it goes to the code at the end of the function...but that doesn't happen because I overlooked the outer loop over `returns`...well spotted.